### PR TITLE
[network] Simplify Network Healthcheck

### PR DESCRIPTION
**Pull request by gregnazario**
### Overview
The network healtcheck was randomly picking a peer to healthcheck.  If
there are a large number of peers, the likelihood of healthchecking a
peer at a regular rate diminishes.  It was as strange behavior that
really doesn't healthcheck at a regular rate.  This now, healthchecks
all peers at a regular rate, creating different nonces for each.  The
nonces are to prevent spoofs of health check responses.

### Why do this now?
It was actually just bothering me to explain that our health checks are non-deterministic.  Now, it's more akin to a regular health check, where pings go out at a regular rate, and are kept track of for each peer that it's healthchecking.

### But wait, won't this create a lot of extra traffic?
Yes, but with the changes coming up for connection limits, we should have a limited amount of healthchecks.  This will prevent us basically causing excess load just from a large number of connections connected to a peer.

### CODE REVIEW

File: network/src/protocols/health_checker/mod.rs [hunk 1]
```diff
@@ -194,72 +194,72 @@ where
...
-                        None => {
-                            trace!(
-                                NetworkSchema::new(&self.network_context),
-                                round = self.round,
-                                "{} No connected peer to ping round: {}",
-                                self.network_context,
-                                self.round
-                            );
-                        }
+                    if peers.is_empty() {
```
Comment (phlip9) [inline network/src/protocols/health_checker/mod.rs:233]: can just `if self.connected.is_empty() {`

File: network/src/protocols/health_checker/mod.rs [hunk 2]
```diff
@@ -194,72 +194,72 @@ where
...
+                        trace!(
+                            NetworkSchema::new(&self.network_context),
+                            round = self.round,
+                            "{} No connected peer to ping round: {}",
+                            self.network_context,
+                            self.round
+                        );
+                    }
+
+                    for peer_id in peers {
```
Comment (phlip9) [inline network/src/protocols/health_checker/mod.rs:243]: `for peer_id in self.connected.keys() {`

**Review by phlip9 — APPROVED** [review]
LGTM (+/- some nits)!

Maybe let's also reduce the default tick rate a little? Right now it's at 1 "tick" per second, so we make 1 outbound ping per second and expect about 1 inbound ping per second. Currently this diff will make ~#V outbound pings/sec and expect #V inbound pings/sec so like 2 * 25 = 50 health checker msg/sec total. Maybe we instead target ~10/sec, so change the `PING_INTERVAL_MS` to 5000.

**Comment by davidiw** [comment]
Likely a net improvement. Let's check a canary before we land this and validate your sentiment... I think we should also pull in all incoming messages into the health checker as indications of liveness... can you explore how "hard" that would be to do in Rust? I think the other concern is that what happens if we have a toooon of FNs can we argue that the weight of this would be less than a state sync service at steady state?

/canary

**Comment by gregnazario** [comment]
> Likely a net improvement. Let's check a canary before we land this and validate your sentiment... I think we should also pull in all incoming messages into the health checker as indications of liveness... can you explore how "hard" that would be to do in Rust? I think the other concern is that what happens if we have a toooon of FNs can we argue that the weight of this would be less than a state sync service at steady state?

Sure, I can see if we can just use other messages as health checks.

The interesting thing here, is we use these nonces in the health check messages to ensure that health check messages aren't spoofed on the message back, except if we're already connected, I'm not exactly sure where the nonce adds value.

I know if we'll have a ton of FNs we'll have a large number of healthchecks, but they are much lighter weight than anything else we currently send.

**Comment by gregnazario** [comment]
@davidiw Can't run the canary due to docker limits.  We'll have to run this ourselves on cluster test for testing

File: network/src/protocols/health_checker/mod.rs (hunk 2 above)
Comment (gregnazario) [reply network/src/protocols/health_checker/mod.rs:244]: Without some rust magic (that i'm not quite aware of yet), I can't do that.

`self.connected.keys()` is an immutable borrow, then `self.nonce()` is a mutable borrow.  So, I'm leaving it with the cloning for now.

**Pull request closed without merging**